### DISCUSSION TITLE
Traffic generator performace issue

### DISCUSSION THREAD

**Discussion by lazna** [dannagle/PacketSender#326] [General]
Starting traffic generator

`packetsender.com --rate %trafgenrate% %trafgenhost% %trafgenport% "%trafgendata%"`

but it generate only about 25Mbps (I5, 2,9Ghz CPU) and it uses only one CPU core. Would it be possible to 'optimize' generating engine, and/or run function on all available CPU cores simultaneously? It only massively sending packets out, so I dont see any synchronizations issues here....

**Comment by dannagle** [comment]
I am happy to report that fix is already in the development branch. It does exactly as you suggest... does an optimization based on how many cores you have available. It then spawns threads accordingly.

What has taken so long is that each thread uses a different source port. I was attempting to make them the same source port and then abandoned that idea. I've decided that the improved speed is worth the trade off of seeing multiple source ports in the logs. 

This fix will be in the next release.

**Reply by lazna** [reply]
Just download and test ver 8.5.2 but performance seems same, checked release notes but found nothing related to this issue. Is improovement already part of this release?

**Reply by dannagle** [reply]
The performance improvement is only if you let it run at max speed.

**Comment by lazna** [comment]
Thanks for notice me about this, but it behave strange. See screenshoot please, all three was taken in first minute of traffic generator with --max option started (fresh rebooted system, no other SW running). It takes a _LOT_ of memory and it stop sending traffic randomly. When press CTRL+C, program itself terminate, but memory remain occupied. Than want to reboot windows (to clear-up memmory), it accept to click to reboot, but then it did not react. After a minute I have to press reset button.

Seems traffic generation algorithm is not in good state and some code review may help. Please check attached pictures
![tg1](https://github.com/dannagle/PacketSender/assets/5476370/534dc98c-6ef9-4fdf-813d-58dd9b3ad9fc)
![tg2](https://github.com/dannagle/PacketSender/assets/5476370/caca9a36-489e-4390-8418-63d5cf11dc62)
![tg3](https://github.com/dannagle/PacketSender/assets/5476370/f5940f0d-2670-466e-a1f0-cf05f134f36b)
